### ISSUE TITLE
Group Kiota PHP repositories

### ISSUE THREAD

**Issue by andrueastman** [microsoft/kiota-abstractions-php#156]
Related to https://github.com/microsoft/kiota/issues/4636 and https://github.com/microsoft/kiota-abstractions-dotnet/issues/238

As a language specific concern, we should look into the feasibility of grouping the repos and outline if this is not feasible in this issue.

**Comment by baywet** [comment]
Whenever this is completed we should also:
- update the quickstart samples https://github.com/microsoft/kiota-samples/pull/6457
- update the quickstart documenation https://github.com/MicrosoftDocs/openapi-docs/pull/110

**Comment by Ndiritu** [comment]
(Subject to future edits as investigation continues, but edits will be listed here)

Composer (dependency manager) supports linking between sub-projects, however Packagist (dependency repository) doesn't support deployments from sub-packages. Packagist expects a package to be hosted within a GitHub repository whose root folder contains a composer.json file. [Reference](https://github.com/composer/packagist/issues/472)

Currently, package deployment works like this, our PHP repos have a Packagist GitHub app set up & a webhook that triggers publishing to Packagist when a tag is created. The version of the packages is determined by the tag or adding a "version" property to the composer.json.

With a mono-repo, our limitation would be in the deployment process.

Various tools exist that split the mono-repo into repositories during deployment & pushes commits & tags to the new repos so that deployment happens.

Two major tools seem viable for this:
### [Mono-repo builder](https://packagist.org/packages/symplify/monorepo-builder) 
which provides various CLI utilities to validate our dependency versions across sub-projects & release automations which we shouldn't need courtesy of release-please. It does the repo split using a release workflow with the `monorepo-split-github-action` & providing it with `repo:write` permissions to split the repository and publish. Mono-repo builder has ~3M+ installs & seems stable at v11.x

### [Splitsh-lite](https://github.com/splitsh/lite) 
which does only the repo subtree split and has been used by [Google Cloud PHP](https://github.com/googleapis/google-cloud-php) mono-repo to deploy their various service libraries as separate packages [ref](https://github.com/googleapis/google-cloud-php/blob/main/.github/workflows/release.yml). They augment this with their own custom scripts.

Mono-repo builder seems most viable however it still leaves us potentially with our individual repositories for abstractions, http etc but we can manage everything via a kiota-php repository.

Pending:
- [ ] Prototype using Mono-repo builder to find any further pitfalls

**Comment by Ndiritu** [comment]
Open to further feedback here @andrueastman, @baywet, @shemogumbe

**Comment by baywet** [comment]
Thank you for the additional information.

So if I'm following things properly, in those scenarios we'd:
- create a mono-repo.
- migrate the "live" version of the code there.
- setup some kind of automation to replicate onto existing repos. (so that packagist webhook is triggered, and the platform has "something to refer to")

Or am I missing anything here?

**Comment by Ndiritu** [comment]
Yes, create a mono-repo. We could keep the current repos for the core libraries for to re-use the current Packagist webhook configs. 
I will confirm all details once I run a PoC within the week.

**Comment by baywet** [comment]
Thanks! The two things I'd want us to get clarity on before we make any decision are:
- Is there a chance that the "packages from the monorepo" could appear in packagist by mistake? that people could take a dependency on them?
- Would "deleting" the existing repos be a breaking change to people who already have a reference to the packages on packagist? (I'm fairly confident the answer to this one is yes)

**Comment by Ndiritu** [comment]
> Is there a chance that the "packages from the monorepo" could appear in packagist by mistake? that people could take a dependency on them?

No. We will not be deploying from the mono-repo and we do not need to set up the Packagist GitHub hook/app on the mono-repo.
We can tag and do a release on GitHub with a CHANGELOG to track changes but these won't get to Packagist without the config described.

> Would "deleting" the existing repos be a breaking change to people who already have a reference to the packages on packagist? (I'm fairly confident the answer to this one is yes)

For this to work, we won't be deleting existing repos. We'll retain the various repos and re-use their existing Packagist configurations so that once changes are tagged on the mono-repo, we push the changes of each sub-project to the corresponding existing GitHub repo with a tag that will trigger an update on Packagist.

We could disable issues & PRs to the existing Kiota core PHP repos, dependabot etc and update the README to point customers to the mono-repo.

**Comment by Ndiritu** [comment]
If there are no further concerns/questions @baywet, we can create the kiota-php mono-repo & I can start making the changes.

**Comment by Ndiritu** [comment]
Realized we can use `git subtree split` to separate the changes in the individual folders and sync the changes to the individual repos via git using the `contents:read` permissions - [buggy sample](https://github.com/sample-monorepo/sample-monorepo/blob/main/.github/workflows/split-monorepo.yml) but illustrates this

**Comment by Ndiritu** [comment]
Running into some trouble pushing to the individual repositories due to access issues. The GitHub token doesn't have scopes to push across repos in an org. Looking for a GitHub app that might be able to do this.

**Comment by baywet** [comment]
Thanks for the deep dive here!
I did have a follow up question: why would we use subtrees instead of submodules?

**Comment by Ndiritu** [comment]
Submodules would mean that each change to a sub-folder is first pushed to the sub-project's remote & merged to the sub-project's main branch... meaning we'd still be reviewing PRs on the sub-projects. Probably cleaner when the sub-project is a dependency - pulling latest changes is easier. Here's a sample monorepo with submodule [setup](https://github.com/sample-monorepo/monorepo-submodules)

Subtrees pull the history of the sub-project making the monorepo the new source of truth. The individual repos would be only for release purposes. Cleaner to make changes

**Comment by baywet** [comment]
Right, this is not an aggregation case but rather a "publishing" case. We want the source of truth to be on the main repo. Sorry about the confusion here for a second.

Thank you for this deep investigation here! (I just saw you created an org & co)

I'm happy with the solution, @andrueastman ?

**Comment by andrueastman** [comment]
Thanks for this as well @Ndiritu 

I'm happy with this as well. Just one question though on this.

> Add automation to parent repos to auto-create PR on push to some release branch pattern - [sample diff](https://github.com/sample-monorepo/sample-abstractions/compare/main...release/abstractions)

Will this mean that multiple PRs are created on each of the existing repo? If so, would it make sense to have these repos only writable by the automation(so that it just writes the source and tag directly to the `main` branch) so that we have less overhead of approving/merging the PRs?

**Comment by Ndiritu** [comment]
@andrueastman I think that would be more ideal but I think @baywet had mentioned that there may be a requirement to have PRs as mandatory on our repos. Could you please clarify?

Also, when ready, could you please create the `kiota-php` repository?

**Comment by baywet** [comment]
The security teams are eventually going to roll out branch policies across all repositories mandating the use of pull requests to merge anything to the main branches. The timelines are not clear at this point, but the last time I heard of it, they were talking about this quarter.

**Comment by baywet** [comment]
Repo created https://github.com/microsoft/kiota-php/

**Comment by Ndiritu** [comment]
@baywet when you can, could you please create a `kiota-bundle-php` repo?
